net_sink.go

```diff
@@ -20,85 +20,117 @@ const (
 	f(sink)
 }
 
 // WithStatsdHost sets the host of the statsd sink otherwise the host is
 // read from the environment variable "STATSD_HOST".
 func WithStatsdHost(host string) SinkOption {
-	return sinkOptionFunc(func(sink *tcpStatsdSink) {
-		sink.statsdHost = host
+	return sinkOptionFunc(func(sink *netSink) {
+		sink.conf.StatsdHost = host
+	})
+}
+
+// WithStatsdProtocol sets the network protocol ("udp" or "tcp") of the statsd
+// sink otherwise the protocol is read from the environment variable
+// "STATSD_PROTOCOL".
+func WithStatsdProtocol(protocol string) SinkOption {
+	return sinkOptionFunc(func(sink *netSink) {
+		sink.conf.StatsdProtocol = protocol
 	})
 }
 
 // WithStatsdPort sets the port of the statsd sink otherwise the port is
 // read from the environment variable "STATSD_PORT".
 func WithStatsdPort(port int) SinkOption {
-	return sinkOptionFunc(func(sink *tcpStatsdSink) {
-		sink.statsdPort = port
+	return sinkOptionFunc(func(sink *netSink) {
+		sink.conf.StatsdPort = port
 	})
 }
 
 // WithLogger configures the sink to use the provided logger otherwise
 // the standard logrus logger is used.
 func WithLogger(log *logger.Logger) SinkOption {
 	// TODO (CEV): use the zap.Logger
-	return sinkOptionFunc(func(sink *tcpStatsdSink) {
+	return sinkOptionFunc(func(sink *netSink) {
 		sink.log = log
 	})
 }
 
-// NewTCPStatsdSink returns a FlushableSink that is backed by a buffered writer
-// and a separate goroutine that flushes those buffers to a statsd connection.
+// NewTCPStatsdSink returns a new NetStink. This function name exists for
+// backwards compatibility.
 func NewTCPStatsdSink(opts ...SinkOption) FlushableSink {
-	outc := make(chan *bytes.Buffer, chanSize) // TODO(btc): parameterize
-	writer := sinkWriter{
-		outc: outc,
-	}
-	// TODO (CEV): this auto loading from the env is bad and should be removed.
-	conf := GetSettings()
-	s := &tcpStatsdSink{
-		outc: outc,
-		// TODO(btc): parameterize size
-		bufWriter: bufio.NewWriterSize(&writer, defaultBufferSize),
+	return NewNetSink(opts...)
+}
+
+// NewNetSink returns a FlushableSink that writes to a statsd sink over the
+// network. By default settings are taken from the enviornment, but can be
+// overridden via SinkOptions.
+func NewNetSink(opts ...SinkOption) FlushableSink {
+	s := &netSink{
 		// arbitrarily buffered
 		doFlush: make(chan chan struct{}, 8),
+
 		// CEV: default to the standard logger to match the legacy implementation.
-		log:        logger.StandardLogger(),
-		statsdHost: conf.StatsdHost,
-		statsdPort: conf.StatsdPort,
+		log: logger.StandardLogger(),
+
+		// TODO (CEV): auto loading from the env is bad and should be removed.
+		conf: GetSettings(),
 	}
 	for _, opt := range opts {
 		opt.apply(s)
 	}
+
+	// Calculate buffer size based on protocol, for UDP we want to pick a
+	// buffer size that will prevent datagram fragmentation.
+	var bufSize int
+	switch s.conf.StatsdProtocol {
+	case "udp", "udp4", "udp6":
```

Comment:
constantize, please :)

Reply:
I normally would, but I think having something like:
```go
const (
	ProtocolUDP  = "udp"
	ProtocolUDP4 = "udp4"
	ProtocolUDP6 = "udp6"
	ProtocolTCP  = "tcp"
	ProtocolTCP4 = "tcp4"
	ProtocolTCP6 = "tcp6"
)
```
Would add more noise than it's worth and wouldn't be as clear.